File: tests/e2e/cases/06-int-ops/byte_swap_u32/byte_swap_u32_verified.cpp
/*
DEMO: 32-bit byte-order reversal via shift+mask (big <-> little endian).

The shift+mask formula reorders the four bytes (b3 b2 b1 b0) into
(b0 b1 b2 b3). SAW proves bit-for-bit equivalence with the Cryptol
`split + reverse + #` reference.

Coverage: logical shifts (`>>`, `<<`) and bitwise OR over `uint32_t` —
not directly exercised by earlier tests.

Expected verdict: VERIFIED.
*/

#include <cstdint>

uint32_t byte_swap_u32(uint32_t x) {
    return ((x >> 24) & 0x000000FFu)
         | ((x >> 8)  & 0x0000FF00u)
         | ((x << 8)  & 0x00FF0000u)
         | ((x << 24) & 0xFF000000u);
}

File: tests/e2e/cases/06-int-ops/byte_swap_u32/byte_swap_u32_verified.rs
// DEMO: 32-bit byte-order reversal (Rust mirror of byte_swap_u32_verified.cpp).
// Same shift+mask formula as the C++ version.

fn byte_swap_u32(x: u32) -> u32 {
    ((x >> 24) & 0x0000_00FF)
        | ((x >> 8) & 0x0000_FF00)
        | ((x << 8) & 0x00FF_0000)
        | ((x << 24) & 0xFF00_0000)
}
